test/TestProjects/Models-Typespec/src/Generated/Models/OutputBaseModelWithDiscriminator.cs

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// <auto-generated/>
+
+#nullable disable
+
+namespace ModelsInCadl.Models
+{
+    /// <summary>
+    /// Output model with a discriminator
+    /// Please note <see cref="OutputBaseModelWithDiscriminator"/> is the base class. According to the scenario, a derived class of the base class might need to be assigned here, or this property needs to be casted to one of the possible derived classes.
+    /// The available derived classes include <see cref="FirstDerivedOutputModel"/> and <see cref="SecondDerivedOutputModel"/>.
+    /// </summary>
+    public abstract partial class OutputBaseModelWithDiscriminator
+    {
+        /// <summary> Initializes a new instance of OutputBaseModelWithDiscriminator. </summary>
+        protected OutputBaseModelWithDiscriminator()
+        {
+        }
+
+        /// <summary> Initializes a new instance of OutputBaseModelWithDiscriminator. </summary>
+        /// <param name="kind"> Discriminator. </param>
+        internal OutputBaseModelWithDiscriminator(string kind)
+        {
+            Kind = kind;
+        }
+
+        /// <summary> Discriminator. </summary>
+        internal string Kind { get; set; }
```

Comment:
From the issue you linked, seems we should generate the code as below.  Even this is for internal usage, do we allow change that value later? 

internal string Kind { get; init; }

Reply:
Pretty sure it should be `init` and not `set`.

Reply:
@cataggar 
If we do that, we will also change every discriminator in mgmt plane as well
And the generator will write `public string Discriminator { get; protected internal set; }` when the discriminator is changed to public by the config
Therefore the setter will always be internal which means the SDK itself controls when to set the discriminator value and I believe it is safe.
I kind of do not want to make a massive change in our generator, and `init` might be depending a specific langver, we just got [an issue about langver](https://github.com/Azure/autorest.csharp/issues/3433), therefore I prefer we keep using `set` here instead of `init`.

Reply:
`init` is [C# 9 from .NET 5](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/init). That issue is about using [UTF-8 string literals](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/utf8-string-literals) from C# 11 from .NET 7 and not being able to compile for .NET 6. `init` should be safe to use.

Reply:
I would buy in that `init` makes more sense and it does not have much risk to change to that, but in the meantime, I will not do it in this PR, because `set` works as well.
We are on the way to eliminate the divergence in our code generator between mgmt, HLC and DPG, therefore we should not introduce another divergence into it. If we want to change it, we should change it for all models in our SDKs.
I will open an issue for this so that we could do it later.

Reply:
Issue created: https://github.com/Azure/autorest.csharp/issues/3452